src/main/java/org/folio/innreach/batch/contribution/service/OngoingContributionEventProcessor.java

```diff
@@ -0,0 +1,94 @@
+import org.folio.innreach.domain.service.ContributionActionService;
+import org.folio.innreach.domain.service.InnReachTransactionActionService;
+import org.folio.innreach.domain.service.impl.TenantScopedExecutionService;
+import org.folio.innreach.external.exception.InnReachConnectionException;
+import org.folio.innreach.external.exception.InnReachGatewayException;
+import org.folio.innreach.external.exception.RetryException;
+import org.folio.innreach.external.exception.ServiceSuspendedException;
+import org.folio.innreach.external.exception.SocketTimeOutExceptionWrapper;
+import org.folio.innreach.util.JsonHelper;
+import org.springframework.beans.factory.annotation.Value;
+import org.springframework.scheduling.annotation.Async;
+import org.springframework.stereotype.Service;
+import org.folio.innreach.dto.Item;
+
+import static org.folio.innreach.domain.entity.ContributionStatus.FAILED;
+import static org.folio.innreach.domain.entity.ContributionStatus.RETRY;
+import static org.folio.innreach.util.InnReachConstants.RETRY_LIMIT_MESSAGE;
+import static org.folio.innreach.util.InnReachConstants.UNKNOWN_EVENT_NAME_MESSAGE;
+import static org.folio.innreach.util.InnReachConstants.UNKNOWN_TYPE_MESSAGE;
+
+@Service
+@RequiredArgsConstructor
+@Log4j2
+public class OngoingContributionEventProcessor {
+
+  private final ContributionActionService contributionActionService;
+  private final InnReachTransactionActionService transactionActionService;
+  private final JsonHelper jsonHelper;
+  private final OngoingContributionStatusService ongoingContributionStatusService;
+  private final TenantScopedExecutionService executionService;
+  @Value("${contribution.retry-attempts}")
+  private int maxRetryAttempts;
+
+  @Async("ongoingSchedulerTaskExecutor")
+  public void processOngoingContribution(OngoingContributionStatus ongoingContributionStatus) {
+    try {
+      log.info("processOngoingContribution:: Processing ongoing contribution event with id {} , tenant {}",
+        ongoingContributionStatus.getId(), ongoingContributionStatus.getTenant());
+      executionService.executeAsyncTenantScoped(ongoingContributionStatus.getTenant(),
+        () -> {
+          checkRetryLimit(ongoingContributionStatus);
+          switch (ongoingContributionStatus.getDomainEventName()) {
+            case ITEM -> processItem(ongoingContributionStatus);
+            case HOLDINGS -> processHoldings(ongoingContributionStatus);
+            case INSTANCE -> processInstance(ongoingContributionStatus);
+            default ->
+              ongoingContributionStatusService.updateOngoingContribution(ongoingContributionStatus, UNKNOWN_EVENT_NAME_MESSAGE, FAILED);
+          }
+        });
+    } catch (ServiceSuspendedException | InnReachConnectionException |
+             SocketTimeOutExceptionWrapper | InnReachGatewayException ex) {
+      ongoingContributionStatusService.updateOngoingContribution(ongoingContributionStatus, RETRY);
```

Comment:
Should we log these exceptions? I think yes. Even if it is `ServiceSuspendendException` we would want to know when and how many times it occurred. Same for other types.

Reply:
I would add the job id to the log messages and any other helpful info.